### PR TITLE
GL-573 | acli migrate pipelines command.

### PR DESCRIPTION
**Pull request by NetraPK**
**Motivation**

Fixes #NNN

**Proposed changes**

**Alternatives considered**

**Testing steps**

1. Follow the [contribution guide](https://github.com/acquia/cli/blob/master/CONTRIBUTING.md#building-and-testing) to set up your development environment.
2. Clear the kernel cache to pick up new and changed commands: `./bin/acli ckc`
3. (add specific steps for this pr)

**Merge requirements**
- [ ] _Bug_, _enhancement_, or _breaking change_ label applied
- [ ] Manual testing by a reviewer

### CODE REVIEW

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 1]
```diff
@@ -0,0 +1,612 @@
...
+
+  /**
+   * @var string
+   */
+  private $gitlabHost;
+
+  /**
+   * @var string
+   */
+  private $acquia_pipeline_file_parse;
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:51]: Class properties should be camel case, local variables should be snake case.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+
+  /**
+   * @var array
+   */
+  private $emptyArray;
+
+  /**
+   * @var array
+   */
+  private $avlChoice = ['yes', 'no'];
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:61]: Let's be very declarative with naming. E.g., "$availableChoices".

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 3]
```diff
@@ -0,0 +1,612 @@
...
+
+  /**
+   * @param \Symfony\Component\Console\Input\InputInterface $input
+   * @param \Symfony\Component\Console\Output\OutputInterface $output
+   *
+   * @return int
+   * @throws \Exception
+   */
+  protected function execute(InputInterface $input, OutputInterface $output) {
+    $this->emptyArray=[];
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:95]: This is a very vague property name. What's its purpose? Name it based on its purpose.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 4]
```diff
@@ -0,0 +1,612 @@
...
+    $this->gitlabHost = $this->getGitLabHost();
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:96]: Notice that in the method, we use camel case for GitLab. Do the same for properties.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 5]
```diff
@@ -0,0 +1,612 @@
...
+    $this->env = $this->validateEnvironment();
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:97]: Don't abbreviate.

Comment (grasmash) [reply src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:97]: Also, this isn't defined as a property? Don't dynamically define properties.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 6]
```diff
@@ -0,0 +1,612 @@
...
+    $this->appUuid = $this->determineCloudApplication();
+    // Get Cloud application.
+    $cloud_application = $this->getCloudApplication($this->appUuid);
+    $this->gitLabProjectDescription = "Source repository for Acquia Cloud Platform application <comment>$this->appUuid</comment>";
+    $project = $this->determineGitLabProject($cloud_application);
+
+    // Migrate acquia-pipeline file
+    $this->getGitLabCiCdVariables($project, $this->appUuid, $cloud_key, $cloud_secret);
+    $this->checkPipelineExists($project);
+    $this->migrateCSStandardTemplate();
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:148]: migrateToCodeStudioStandardTemplate()

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 7]
```diff
@@ -0,0 +1,612 @@
...
+    $this->migrateVariables();
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:149]: migrateVariablesSection()

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 8]
```diff
@@ -0,0 +1,612 @@
...
+    $this->migrateBuild();
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:150]: migrateBuildSection()

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 9]
```diff
@@ -0,0 +1,612 @@
...
+    $this->migratePostDeploy();
+    $this->io->success([
+      "",
+      "*********** Migration completed successfully. *********\n",
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:154]: You don't need `\n` here, new lines are added automatically.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 10]
```diff
@@ -0,0 +1,612 @@
...
+   *
+   * @return array
+   */
+  protected function determineGitLabProject(ApplicationResponse $cloud_application) {
+    $projects = $this->gitLabClient->projects()->all(['search' => $cloud_application->uuid]);
+    if (count($projects) == 1) {
+      return reset($projects);
+    }
+    else {
+      throw new AcquiaCliException("[ERROR] '{$cloud_application->name}' is not configured as Code Studio project.\n [Help] You can configure the '{$cloud_application->name}' project in Code Studio by using 'acli cs:wizard' command \n You can find Code Studio documentation at https://docs.acquia.com/code-studio/getting-started/#create-code-studio-project \n");
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:198]: We don't add formatting to exception messages.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 11]
```diff
@@ -0,0 +1,612 @@
...
+    $GLAB_TOKEN_NAME = "ACQUIA_GLAB_TOKEN_SECRET";
+    $gitlab_cicd_existing_variables = $this->gitLabClient->projects()->variables($project['id']);
+    foreach ($gitlab_cicd_existing_variables as $variable) {
+      if ($variable['key'] == $GLAB_TOKEN_NAME) {
+        return $variable;
+      }
+    }
+    foreach ($gitlab_cicd_existing_variables as $variable) {
+      if ($variable['key'] != $GLAB_TOKEN_NAME) {
+        throw new AcquiaCliException("[Error] Code Studio CI/CD environment 'Variable_Name' variable is not configured properly. \n[Help] To set all the CI/CD environment variables, configure the project in Code Studio by using 'acli cs:wizard' command \n You can find Code Studio documentation at https://docs.acquia.com/code-studio/getting-started/#create-code-studio-project \n OR \n Set the above required CI/CD environment variables in Code Studio project.\n");
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:222]: Don't format this message.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 12]
```diff
@@ -0,0 +1,612 @@
...
+  /**
+   * Check acquia-pipeline.yml file exists in the root repo and remove ci_config_path from codestudio project.
+   * @param array $project
+   * @throws \Acquia\Cli\Exception\AcquiaCliException
+   */
+  protected function checkPipelineExists(array $project) {
+    $pipelines_filepath_yml = Path::join($this->repoRoot, 'acquia-pipelines.yml');
+    $pipelines_filepath_yaml = Path::join($this->repoRoot, 'acquia-pipelines.yaml');
+    if ($this->localMachineHelper->getFilesystem()->exists($pipelines_filepath_yml) or $this->localMachineHelper->getFilesystem()->exists($pipelines_filepath_yaml)) {
+      // print("file exists");
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:238]: Remove this debugging statements.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 13]
```diff
@@ -0,0 +1,612 @@
...
+    $auto_devops_pipeline = [
+        'include' => ['project' => 'acquia/standard-template', 'file' => '/gitlab-ci/Auto-DevOps.acquia.gitlab-ci.yml'],
+      ];
+    $auto_devops_dump_file = Yaml::dump($auto_devops_pipeline);
+    $auto_devops_parse_file = Yaml::parse($auto_devops_dump_file);
+    $this->emptyArray = array_merge($this->emptyArray, $auto_devops_parse_file);
+    $this->createGitlabCiFile($this->emptyArray);
+  }
+
+  //todo : catch execpetion if any and throw error.
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:266]: Address this.

**Comment by grasmash** [comment]
Please address the PHPStan errors here: https://github.com/acquia/cli/runs/6558025604?check_suite_focus=true

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 14]
```diff
@@ -0,0 +1,612 @@
...
+
+  /**
+   * Migrating varibales to .gitlab-ci.yml file.
+   */
+  protected function migrateVariables() {
+    $varData = isset($this->acquia_pipeline_file_parse['variables'])?$this->acquia_pipeline_file_parse['variables']:FALSE;
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:272]: This is probably going to fail PHPCS for a few reasons.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 15]
```diff
@@ -0,0 +1,612 @@
...
+    if ($varData){
+      $variables_dump = Yaml::dump(['variables' => $varData]);
+      $remove_global = preg_replace('/global:/', '', $variables_dump);
+      $variables_parse = Yaml::parse($remove_global);
+      $this->emptyArray = array_merge($this->emptyArray, $variables_parse);
+      $this->createGitlabCiFile($this->emptyArray);
+      $this->io->writeln([
+        "",
+        "\n*********** Migration completed for variables section. *********\n",
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:281]: Remove the new lines and the ****. If you're going to add styling, it needs to be codified and consistently. The `****` is a one-off.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 16]
```diff
@@ -0,0 +1,612 @@
...
+  /**
+   * Migrating build job to .gitlab-ci.yml file.
+   */
+  protected function migrateBuild() {
+    $buildJob=isset($this->acquia_pipeline_file_parse['events']['build']['steps'])?$this->acquia_pipeline_file_parse['events']['build']['steps']:[];
+    if (!empty($buildJob)){
+      $response_composer = NULL;
+      $response_BLT = NULL;
+
+      $chunck = array_chunk($buildJob, 1);
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:301]: I think you can remove this and just iterate directly on $build_job.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 17]
```diff
@@ -0,0 +1,612 @@
...
+      foreach ($chunck as $key => $value){
+        $keysvar = array_keys($value[0])[0];
+        if(empty($value[0][$keysvar]['script'])){
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:304]: You probably also need to check that this array key exists here. E.g., `array_key_exists('script', $value[0][$keysvar])`

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 18]
```diff
@@ -0,0 +1,612 @@
...
+          continue;
+        }
+        foreach ($value[0][$keysvar]['script'] as $script){
+          if(strstr ($script, 'composer')) {
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:308]: It doesn't seem enough to just check for this. What it they are running `composer validate` or `composer run-script`? It needs to actually be `composer install`.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 19]
```diff
@@ -0,0 +1,612 @@
...
+            if(empty($response_composer)){
+              $ques = "Composer script is part of Code Studio Auto DevOps Pipeline, do you still want to migrate Composer script?(yes,no)";
+              $response_composer = $this->getCustResponse($ques, $this->avlChoice, $this->defaultChoice);
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:311]: You should change this to `$this->io->confirm` to make it a yes/no.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 20]
```diff
@@ -0,0 +1,612 @@
...
+            if($response_composer == 'yes'){
+              $arrayempty[$keysvar]['script'][] = $script;
+              continue;
+            }
+            // print_r("Using job from code studio\n\n");
+          }
+          else if(strstr ($script, '${BLT_DIR}')) {
+            if(empty($response_BLT)){
+              $ques = "BLT script is part of Code Studio Auto DevOps Pipeline, do you still want to migrate BLT script?(yes,no)";
+              $response_BLT = $this->getCustResponse($ques, $this->avlChoice, $this->defaultChoice);
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:322]: You should change this to $this->io->confirm() to make it a yes/no.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 21]
```diff
@@ -0,0 +1,612 @@
...
+      $this->setGitLabClient($gitlab_client);
+    }
+    return $this->gitLabClient;
+  }
+
+  /**
+   * @param Client $client
+   */
+  public function setGitLabClient($client) {
+    $this->gitLabClient = $client;
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:609]: This looks like duplicate code.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 22]
```diff
@@ -0,0 +1,612 @@
...
+
+  /**
+   * Getting the gitlab token from user.
+   * @param string $gitlab_host
+   *
+   * @return string
+   * @throws \Acquia\Cli\Exception\AcquiaCliException
+   */
+  protected function getGitLabToken(string $gitlab_host): string {
+    if ($this->input->getOption('gitlab-token')) {
```
Comment (grasmash) [inline src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:503]: This looks like duplicate code.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php (hunk 2 above)
Comment (NetraPK) [reply src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:61]: Done.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php (hunk 1 above)
Comment (NetraPK) [reply src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:51]: Done.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 23]
```diff
@@ -0,0 +1,612 @@
...
+  /**
+   * @var string
+   */
+  private $defaultChoice = 'no';
+
+  /**
+   * {inheritdoc}.
+   */
+  protected function configure() {
+    $this->setDescription('Migrate acquia-pipeline file to gitlab file for a given Acquia Cloud application')
```
[the remark this answers is not in the document]
Comment (NetraPK) [reply src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:77]: Done.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php (hunk 10 above)
Comment (NetraPK) [reply src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:205]: I wanted to add help along with error message then how should I add it?

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php (hunk 14 above)
Comment (NetraPK) [reply src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php]: I tested it PHPCS is passing.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php (hunk 18 above)
Comment (NetraPK) [reply src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:313]: As in CodeStudio Standard template we have composer validate and composer install so I added composer string to check.

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php [hunk 24]
```diff
@@ -0,0 +1,612 @@
...
+      $chunck = array_chunk($buildJob, 1);
+      foreach ($chunck as $key => $value){
+        $keysvar = array_keys($value[0])[0];
+        if(empty($value[0][$keysvar]['script'])){
+          continue;
+        }
+        foreach ($value[0][$keysvar]['script'] as $script){
+          if(strstr ($script, 'composer')) {
+            if(empty($response_composer)){
+              $ques = "Composer script is part of Code Studio Auto DevOps Pipeline, do you still want to migrate Composer script?(yes,no)";
```
[the remark this answers is not in the document]
Comment (NetraPK) [reply src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:315]: Can you please provide me one example?

File: src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php (hunk 10 above)
Comment (grasmash) [reply src/Command/CodeStudio/CodeStudioPipelinesMigrateCommand.php:204]: You'd need to catch the exception here: https://github.com/acquia/cli/blob/master/src/EventListener/ExceptionListener.php#L54

**Pull request closed without merging**